[ModuleInterface] Trap when trying to print a null type

jrose-apple opened the pull request:
We want to catch these bugs sooner rather than later, like the one Harlan fixed in #19175. Trapping deterministically is the best way to do so, and it's better than generating nonsense.

jrose-apple commented:
@swift-ci Please smoke test

harlanhaskins reviewed (approved):
Nitpick: Can you change this to an assert? Just for the message. Other than that, thank you!

jrose-apple commented:
I wanted it to still trap in no-asserts builds, and I didn't want to repeat the check. Which of these three shall we sacrifice? :-)

jrose-apple commented:
That was a serious question! If you think the message is more useful, we can drop one of the other features.

jrose-apple commented:
Sure, I'll go with that.

jrose-apple committed:
[ModuleInterface] Trap when trying to print a null type

We want to catch these bugs sooner rather than later, like the one
Harlan fixed in eb75ad8. Trapping deterministically is the best way to
do so, and it's better than generating nonsense.

jrose-apple commented:
Aha, found an even better solution: `llvm::report_fatal_error`. Admittedly this only works because we have a custom handler that prints backtraces, but that's still good enough.

@swift-ci Please smoke test
